build_tools/cmake/iree_run_module_test.cmake

```diff
@@ -0,0 +1,258 @@
+#       for various reasons, e.g., upstream llvm backend.
+#   UNSUPPORTED_PLATFORMS: Platforms
+#       (android-arm64-v8a|riscv64-Linux|riscv32-Linux) not supported by the
+#       test target. The target will be skipped during crosscompile.
+#   TIMEOUT: (optional) Test timeout.
+#
+# Example:
+#
+# iree_run_module_test(
+#   NAME
+#     person_detect_int8_correctness_test
+#   BENCHMARK_MODULE_SRC
+#     "TFLite/PersonDetect-int8"
+#   EXPECTED_OUTPUT
+#     "1x2xi8=[72 -72]"
+#   UNSUPPORTED_PLATFORMS
+#     "android-arm64-v8a"
+# )
+#
+# iree_run_module_test(
+#   NAME
+#     mobilenet_v1_fp32_correctness_test
+#   BENCHMARK_MODULE_SRC
+#     "TFLite/MobileNetV1-fp32,imagenet"
+#   EXPECTED_OUTPUT
+#     "mobilenet_v1_fp32_expected_output.txt"
+#   UNSUPPORTED_PLATFORMS
+#     "android-arm64-v8a"
+#     "riscv32-Linux"
+# )
+
+function(iree_run_module_test)
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;BENCHMARK_MODULE_SRC;MODULE_SRC;DRIVER;EXPECTED_OUTPUT;TIMEOUT"
+    "RUNNER_ARGS;LABELS;XFAIL;UNSUPPORTED_PLATFORMS"
+    ${ARGN}
+  )
+
+  if(CMAKE_CROSSCOMPILING AND "hostonly" IN_LIST _RULE_LABELS)
+    return()
+  endif()
+
+  get_platform(_PLATFORM)
+  if(_PLATFORM IN_LIST _RULE_UNSUPPORTED_PLATFORMS)
+    return()
+  endif()
+
+  if(_RULE_BENCHMARK_MODULE_SRC AND _RULE_MODULE_SRC)
+    message(SEND_ERROR "Can only specify either MODULE_SRC or BENCHMARK_MODULE_SRC")
+  endif()
+
+  if(NOT _RULE_BENCHMARK_MODULE_SRC AND NOT _RULE_MODULE_SRC)
+    message(SEND_ERROR "Need to specify either MODULE_SRC or BENCHMARK_MODULE_SRC")
+  endif()
+
+  # Benchmark suite needs to be stored at the location of
+  # `IREE_BENCHMARK_SUITE_DIR` or the test target is bypassed.
+  if(_RULE_BENCHMARK_MODULE_SRC AND NOT DEFINED IREE_BENCHMARK_SUITE_DIR)
+    return()
+  endif()
+
+  set(_RUNNER_TARGET "iree-run-module")
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  if(_RULE_MODULE_SRC)
+    get_filename_component(_SRC "${_RULE_MODULE_SRC}" REALPATH)
+  elseif(_RULE_BENCHMARK_MODULE_SRC)
+    set(_MODULE_FLAG_DIR "${IREE_BENCHMARK_SUITE_DIR}/${_RULE_BENCHMARK_MODULE_SRC}/")
+    # Find the platform specific module flag file with matching path name.
+    # TODO(#10391): Update this logic with the new benchmark framework.
+    if(_PLATFORM STREQUAL "riscv64-Linux")
+      set(_FLAGFILE_HINT_PATH "${_MODULE_FLAG_DIR}/iree-llvm-cpu*RV64*__full-inference,default-flags/flagfile")
+    elseif(_PLATFORM STREQUAL "riscv32-Linux")
+     set(_FLAGFILE_HINT_PATH "${_MODULE_FLAG_DIR}/iree-llvm-cpu*RV32*__full-inference,default-flags/flagfile")
+    elseif(_PLATFORM STREQUAL "android-arm64-v8a")
+      set(_FLAGFILE_HINT_PATH "${_MODULE_FLAG_DIR}/iree-llvm-cpu*ARM64-v8A*__big-core,full-inference,default-flags/flagfile")
+    else()  # X86_64
+      set(_FLAGFILE_HINT_PATH "${_MODULE_FLAG_DIR}/iree-llvm-cpu*x86_64*__full-inference,default-flags/flagfile")
+    endif()
```

Comment:
I'm still not a fan of how flagfiles are used here or how this integrates with benchmarks (I'm not even sure how I'd go about extending this for Emscripten or another new platform)

I'd be happier if this was just a generic way to use `iree-run-module` with inputs and expected outputs...

Reply:
The main usage of the flagfiles is to get the filename for the platform-specific vmfb, which is currently coded with the hash. When the benchmark framework is done, we could rework this part, as marked in TODO. Hopefully all the information is rendered in a json file, and we just parse the platform/vmfb information there.

To avoid confusion, I will move all other `iree-run-module` flags into `RUNNER_ARGS` so it becomes obvious to the users instead of relying on the embedded information in the flagfiles.

Reply:
How about we make `iree_run_module_test` a generic thing that just a module and the args to run it with and use a separate wrapper that pulls everything out of the benchmark files. Then we cleanly separate the generic component from the hacky thing that we hope to get rid of. Scott gets something he can use for emscripten, and you are also able to move forward leveraging the already-built artifacts

Reply:
Yes, that's what Scott suggested in the other comment. I will change the implementation to that in the next commit. :)